fdb-record-layer-lucene/src/test/java/com/apple/foundationdb/record/lucene/directory/AgilityContextTest.java

```diff
@@ -0,0 +1,217 @@
+                        return null;
+                    });
+                }
+            }
+            context.commit();
+        }
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    void testAgilityContextConcurrent(boolean useAgile)  throws ExecutionException, InterruptedException {
+        try (FDBRecordContext context = fdb.openContext()) {
+            final AgilityContext agilityContext = AgilityContext.factory(context, useAgile);
+            testAgilityContextConcurrentSingleObject(agilityContext, true);
+            context.commit();
+        }
+        assertLoopThreadsValues();
+    }
+
+    @ParameterizedTest
+    @ValueSource(booleans = {false, true})
+    void testAgilityContextConcurrentNonExplicitCommits(boolean useAgile) throws ExecutionException, InterruptedException {
+        for (int sizeQuota : new int[] {1, 2, 7, 21, 100, 10000}) {
+            final RecordLayerPropertyStorage.Builder insertProps = RecordLayerPropertyStorage.newBuilder()
+                    .addProp(LuceneRecordContextProperties.LUCENE_AGILE_COMMIT_SIZE_QUOTA, sizeQuota);
+
+            try (FDBRecordContext context = openContext(insertProps)) {
+                final AgilityContext agilityContext = AgilityContext.factory(context, useAgile);
+                testAgilityContextConcurrentSingleObject(agilityContext, false);
+                context.commit();
+            }
+        }
+        assertLoopThreadsValues();
+    }
+
+    void testAgilityContextOneLongWrite(int loopCount, int sizeLimit, int timeLimit) {
+        final RecordLayerPropertyStorage.Builder insertProps = RecordLayerPropertyStorage.newBuilder()
+                .addProp(LuceneRecordContextProperties.LUCENE_AGILE_COMMIT_SIZE_QUOTA, sizeLimit)
+                .addProp(LuceneRecordContextProperties.LUCENE_AGILE_COMMIT_TIME_QUOTA, timeLimit);
+
+        String ShelSilverstein =
+                "Oh, if you’re a bird, be an early bird\n" +
+                "And catch the worm for your breakfast plate.\n" +
+                "If you’re a bird, be an early early bird--\n" +
+                "But if you’re a worm, sleep late.";
+        try (FDBRecordContext context = openContext(insertProps)) {
+            final AgilityContext agilityContext = AgilityContext.factory(context, true);
+            for (int i = 0; i < loopCount; i++) {
+                byte[] key = Tuple.from(2023, i).pack();
+                byte[] val = Tuple.from(i, ShelSilverstein).pack();
+                agilityContext.set(key, val);
+            }
+            context.commit();
+        }
+        try (FDBRecordContext context = openContext(insertProps)) {
+            final AgilityContext agilityContext = AgilityContext.factory(context, false);
+            for (int i = 0; i < loopCount; i++) {
+                byte[] key = Tuple.from(2023, i).pack();
+                byte[] val = Tuple.from(i, ShelSilverstein).pack();
+                agilityContext.get(key).thenApply(bytes -> {
```

Comment:
Here also, you don't join the future and won't see failures.

Reply:
✅